### PR TITLE
[fix] Automatically add relevant Jinja methods to cache key if present

### PR DESCRIPTION
**Pull request by john-bodley**
### CATEGORY

Choose one

- [x] Bug Fix
- [ ] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY

The URL parameters were not being used in the cache key resulting in issues per the attached screenshot. This PR adds the URL params (as well as other JInja macros including `current_user_id` and `current_username`) to the cache key by default if they are leveraged by the query. 

#### Before

Screen Shot 2020-04-17 at 11 52 20 AM

#### After 

Screen Shot 2020-04-17 at 11 51 25 AM

### TEST PLAN

Updated unit tests.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### REVIEWERS

to: @villebro

### CODE REVIEW

File: tests/core_tests.py [hunk 1]
```diff
@@ -173,7 +173,6 @@ def test_viz_cache_key(self):
         viz = slc.viz
         qobj = viz.query_obj()
         cache_key = viz.cache_key(qobj)
-        self.assertEqual(cache_key, viz.cache_key(qobj))
```
Comment (john-bodley) [inline tests/core_tests.py:176]: This seems unnecessary, i.e., it's equivalent to, 

```python
self.assertEqual(viz.cache_key(qobj), viz.cache_key(qobj))
```

Comment (villebro) [reply tests/core_tests.py:176]: I have no idea what I was thinking there 😄

**Review by villebro — COMMENTED** [review]
I wonder if there is the risk that this will cause lots of unnecessary cache misses? I would almost prefer to use [`cache_key_wrapper`](https://superset.incubator.apache.org/sqllab.html#superset.jinja_context.CacheKeyWrapper.cache_key_wrapper) for a case like this, as it should be the main tool for adding non-standard cache keys. I can imagine it's a feature that very few probably use or know about, so perhaps it should be promoted somehow when people use jinja in their queries.

**Comment by john-bodley** [comment]
@villebro I wasn't aware of `cache_key_wrapper`. It seems like that method is used primarily for ad-hoc datasource level caching outside of the constructs of Superset, however `url_params` is a construct that Superset supports and can be used outside of the Jinja construct.

**Comment by villebro** [comment]
@john-bodley the way I see it, `url_params` is a slightly special property in chart metadata, and should only be regarded as default values that are overridden if the parameter is present in the URL. It's worth keeping in mind that adding `url_params` to the cache key will invalidate the cache key for every query/chart in a dashboard if a query parameter is added to the URL, even if it is not used by the query. It is for this purpose that `cache_key_wrapper` was added, ie. making sure that any additional variable that needs to be considered in the query is considered in the cache key. Given the fact that using `cache_key_wrapper` can solve this problem without impacting other charts/queries, I would propose using that for now.

**Comment by john-bodley** [comment]
@villebro I agree with your point (and noted) that the presence of a URL parameter does not mean it will be used by the query but per the documentation, 

> As you create a visualization form this SQL Lab query, you can pass parameters in the explore view as well as from the dashboard, and it should carry through to your queries.
> 
> Default values for URL parameters can be defined in chart metadata by adding the key-value pair `url_params: {'foo': 'bar'}` 

I perceive one would expect that if one were using them as defined they wouldn't expect the cached payload to be incorrect, i.e., it feels like a bug if they didn't know they needed to wrap the call with `cache_key_wrapper`. 

Maybe there's a common ground we could agree on (however I wasn't able to get it working but maybe you could shed some light on maybe how to do it). Is there a way of possibly calling `cache_key_wrapper` from within `url_param` to register `url_params` as extra cache keys? That would ensure these are only added to the cache if they're used by the query.

**Comment by villebro** [comment]
I agree, users will probably expect url_params to be added to cache keys. Perhaps we default `url_params` to add the value to `extra_cache_keys`, but leave an option to opt out:

```python
def url_param(param: str, default: Optional[str] = None, add_to_cache_keys: bool=True) -> Optional[Any]:
   ...
```

Then we also need to change the regex `has_calls_to_cche_key_wrapper` in `connectors/models/sqla.py` to be something along the lines of
```python
regex = re.compile(r"\{\{.*(cache_key_wrapper\(.*\)|url_param\(.*).*\}\}")
```
to make sure we don't unnecessarily trigger re-evaluation of jinja code. As I added this feature, I can put together a PR for it if you wish.

Oh, the syntax for `cache_key_wrapper` is fairly simple, just wrap it around whatever JSON serializable value you want to add to keys:
```python
{{ cache_key_wrapper(url_param("key", "my default value")) }}
```

**Comment by villebro** [comment]
Looking more closely at the default jinja functions, it appears they should all default to being added to extra cache keys, except `filter_values` and `form_data`:
- `url_param`
- `current_user_id`
- `current_username`
Then regular users wouldn't need to concern themselves with wrapping the function calls, and `cache_key_wrapper` would be something that mostly more advanced users need when introducing custom jinja functions.

**Comment by john-bodley** [comment]
@villebro thanks for the pointers. I can take a stab at the PR.

**Comment by john-bodley** [comment]
@villebro I took a brief stab at this ([commit](https://github.com/apache/incubator-superset/pull/9572/commits/145b8ea039468aae447a44ed632efdcb6fba5372)) though it's untested. 

You have more context here but my sense is that the `cache_key_wrapper` method may now be redundant. Maybe it would be best if you took over this PR (per your previous suggestion) as you're much more well versed on exactly how this all works.

My one concern is it seems that the `extra_cache_keys` is merely a list of values, i.e., `user_1`, `user_2` etc. however there's not scope and thus there could be collision. I wonder if it should be a dictionary of values keyed by method with more context, i.e,. 

```python
{
    "current_user_id": ["user_1", "user_2"], 
    "url_param": ["foo"],
}
```

**Comment by villebro** [comment]
Thanks, I'll test and review shortly. It would be great if you can see this through, as getting a fresh set of eyes on the old code is always positive. If we can completely drop `cache_key_wrapper` all the merrier.

File: superset/jinja_context.py [hunk 2]
```diff
@@ -122,12 +78,52 @@ def filter_values(column: str, default: Optional[str] = None) -> List[str]:
...
+        r"cache_key_wrapper\(.*\)|"
+        r"url_param\(.*\)"
+        r").*\}\}"
+    )
+
     def __init__(self, extra_cache_keys: Optional[List[Any]] = None):
-        self.extra_cache_keys = extra_cache_keys
+        self.extra_cache_keys = extra_cache_keys or []
+
+    def current_user_id(self, add_to_cache_keys: bool = True) -> Optional[int]:
```
Comment (villebro) [inline superset/jinja_context.py:99]: I think the references to the moved methods need to be updated in `docs/sqllab.rst`.

File: superset/viz.py [hunk 3]
```diff
@@ -395,6 +395,12 @@ def cache_key(self, query_obj, **extra):
         cache_dict["extra_cache_keys"] = self.datasource.get_extra_cache_keys(query_obj)
         cache_dict["rls"] = security_manager.get_rls_ids(self.datasource)
         cache_dict["changed_on"] = self.datasource.changed_on
+
+        url_params = self.form_data.get("url_params")
+
+        if url_params:
+            cache_dict["url_params"] = url_params
+
```
Comment (villebro) [inline superset/viz.py:403]: I don't think this should be added here, as calling `url_params` in the Jinja context will now add the relevant values to `extra_keys`.

File: superset/viz_sip38.py [hunk 4]
```diff
@@ -436,6 +436,12 @@ def cache_key(self, query_obj, **extra):
         cache_dict["extra_cache_keys"] = self.datasource.get_extra_cache_keys(query_obj)
         cache_dict["rls"] = security_manager.get_rls_ids(self.datasource)
         cache_dict["changed_on"] = self.datasource.changed_on
+
+        url_params = self.form_data.get("url_params")
+
+        if url_params:
+            cache_dict["url_params"] = url_params
+
```
Comment (villebro) [inline superset/viz_sip38.py:444]: same here (thanks for keeping this in sync)

File: tests/core_tests.py [hunk 5]
```diff
@@ -206,6 +205,15 @@ def test_cache_key_changes_when_datasource_is_updated(self):
...
 
+    def test_cache_key_changes_with_url_parameters(self):
+        self.login(username="admin")
+        slc = self.get_slice("Girls", db.session)
+        viz = slc.viz
+        qobj = viz.query_obj()
+        cache_key_original = viz.cache_key(qobj)
+        viz.form_data["url_params"] = {"foo": "bar"}
+        self.assertNotEqual(cache_key_original, viz.cache_key(qobj))
+
```
Comment (villebro) [inline tests/core_tests.py:216]: I would recommend adding a test making sure that the cache key does not change when `url_params` changes, unless some part of the `query_obj` is referencing one of the keys present in `url_params` using Jinja. The point being we want to make sure only the url params that are used affect the cache key.

**Review by villebro — CHANGES_REQUESTED** [review]
I think this looks mostly good, just a couple of small changes that I believe are vital for this functionality. Also, we should probably add a note in `UPDATING.md` informing current users of `cache_key_wrapper` that the new default behaviour for `url_param`, `current_username` and `current_user_id` will automatically add the value to `extra_cache_keys` unless explicitly setting `add_to_cache_keys = False`.

**Comment by villebro** [comment]
> @villebro I took a brief stab at this ([commit](https://github.com/apache/incubator-superset/pull/9572/commits/145b8ea039468aae447a44ed632efdcb6fba5372)) though it's untested.
> 
> You have more context here but my sense is that the `cache_key_wrapper` method may now be redundant. Maybe it would be best if you took over this PR (per your previous suggestion) as you're much more well versed on exactly how this all works.

The `cache_key_wrapper` method is good to leave as-is for cases where users want to add values to `extra_cache_keys` other than the predefined `current_username()` etc. However, the documentation probably needs to be updated to reflect the fact that the predefined jinja functions don't need to be wrapped.

> My one concern is it seems that the `extra_cache_keys` is merely a list of values, i.e., `user_1`, `user_2` etc. however there's not scope and thus there could be collision. I wonder if it should be a dictionary of values keyed by method with more context, i.e,.
> 
> ```python
> {
>     "current_user_id": ["user_1", "user_2",], 
>     "url_param": [{"foo": "bar"},],
> }
> ```
> 
> Note in the case of the URL parameters having both the key and value is probably necessary for completeness.

With regards to `extra_cache_keys` being a list, I don't think it's a big issue, as the extra cache keys will probably always be added sequentially. So if we're adding two keys `key1` and `key2`, with values `value1` and `value2` respectively, it probably won't make a big difference if we're using `{"key1": "value1", "key2": "value2"}` or `["value1", "value2"]`. While there is the theoretical risk that the code adds different keys inder different circumstances, to my understanding this has not been an issue during the time this has been available (since last summer).

File: tests/core_tests.py (hunk 5 above)
Comment (john-bodley) [reply tests/core_tests.py:216]: Agreed.

File: superset/viz.py (hunk 3 above)
Comment (john-bodley) [reply superset/viz.py:403]: This needs to be removed.

File: superset/viz_sip38.py (hunk 4 above)
Comment (john-bodley) [reply superset/viz_sip38.py:444]: See ^^.

**Comment by john-bodley** [comment]
@villebro could you think of an example when `cache_key_wrapper` would be used given we're wrapping numerous methods with this? I'm trying to determine whether we still need this method.

**Comment by villebro** [comment]
@john-bodley I think any `Callable` passed to `JINJA_CONTEXT_ADDONS` in `superset_config.py` might require wrapping.

**Comment by john-bodley** [comment]
@villebro thanks for the feedback. I believe this is now ready for re-review.

**Review by villebro — APPROVED** [review]
Very nice! 👍  If you could also add a note in `UPDATING.md` to let users know this is now much simpler, ie. builtin Jinja functions don't need to be wrapped anymore, this is ready to go.

**Comment by john-bodley** [comment]
@villebro I updated `UPDATING.md` per your suggestion.

**Pull request merged**